ehsm_kms_service/secret_manager_apis.js

```diff
@@ -265,9 +360,709 @@ const createSecret = async (res, appid, payload, DB) => {
+    try {
+        //get and check Param in payload
+        const deleteTime = new Date().getTime()
+        let secretName = getParam_String(payload, 'secretName')
+        let recoveryPeriod = payload['recoveryPeriod']
+        let forceDelete = payload['forceDelete']
+        if (!checkStringParam(secretName, true, SECRETNAME_LENGTH_MAX)) {
+            res.send(_result(400, `secretName cannot be empty, must be string and length not more than ${SECRETNAME_LENGTH_MAX}`))
+            return
+        }
+        if (typeof (recoveryPeriod) !== 'number' && recoveryPeriod != '' && recoveryPeriod != undefined) {
+            res.send(_result(400, 'recoveryPeriod must be number'))
+            return
+        }
+        if (recoveryPeriod < 1 || recoveryPeriod > 365) {
+            res.send(_result(400, 'recoveryPeriod must be more then 0 and less than 366'))
+            return
+        }
+        if (forceDelete != '' && forceDelete != undefined) {
+            if (forceDelete != 'true' && forceDelete != 'false') {
+                res.send(_result(400, 'forceDelete must be true or false'))
+                return
+            }
+        }
+        //Set default delete time
+        if (recoveryPeriod == '' || recoveryPeriod == undefined) {
+            recoveryPeriod = DEFAULT_DELETE_RECOVERY_DAYS
+        }
+
+        if (forceDelete == 'true') {
+            //Force delete secret metadata and secret version data by appId and secretName
+            if (await forceDeleteData(DB, appid, secretName)) {
+                res.send(_result(200, `The ${base64_decode(secretName)} delete success`))
+                return
+            } else {
+                res.send(_result(400, 'force delete failed'))
+                return
+            }
+        } else {
+            //change delete secret metadata and secret version data according to the scheduled deletion time
+            //Query and change the secret metadata
+            const query_matadata = {
+                selector: {
+                    appid,
+                    secretName
+                },
+                fields: [
+                    '_id',
+                    '_rev',
+                    'appid',
+                    'secretName',
+                    'encryptionKeyId',
+                    'description',
+                    'createTime',
+                    'deleteTime',
+                    'plannedDeleteTime',
+                    'rotationInterval',
+                    'lastRotationDate',
+                    'nextRotationDate'
+                ],
+                limit: 1
+            }
+            let secret_metadata_res = await DB.partitionedFind('secret_metadata', query_matadata)
+            if (secret_metadata_res.docs.length > 0) {
+                secret_metadata_res.docs[0].deleteTime = deleteTime
+                secret_metadata_res.docs[0].plannedDeleteTime = deleteTime + 24 * 60 * 60 * 1000 * recoveryPeriod
```

Comment:
recoveryPeriod may be 4/3 or pi, so that plannedDeleteTime is not integer as we defined. Here we can round the result to an integer.

Reply:
A Integer verification will be added